### PR TITLE
introduce pasteFn

### PR DESCRIPTION
**Pull request by dvnrsn**

### CODE REVIEW

File: src/rich-text-editor.component.js [hunk 1]
```diff
@@ -27,6 +27,21 @@ export const RichTextEditor = forwardRef((props, editorRef) => {
     }
   }
 
+  function interceptPaste (event) {
+    // https://developer.mozilla.org/en-US/docs/Web/Events/paste
+    event.preventDefault()
+    event.stopPropagation()
+    let paste = (window.clipboardData).getData('text');
```
Comment (dvnrsn) [inline src/rich-text-editor.component.js]: MDN has `e.clipboardData || window.clipboardData` but I found that e.clipboardData would not transfer HTML.

**Comment by dvnrsn** [comment]
Usage: 
```js
function stripHTML(content) {
  ...
}
return (
<RichTextEditor 
  ref={editorRef}
  pasteFn={content => editorRef.current.setHTML(stripHTML(content))
```

File: src/rich-text-editor.component.js [hunk 2]
```diff
@@ -27,6 +27,21 @@ export const RichTextEditor = forwardRef((props, editorRef) => {
...
+    paste = props.pasteFn(paste)
+  }
+
+  useEffect(() => {
+    if (props.pasteFn) {
+      divRef.current.addEventListener('paste', interceptPaste)
+      return divRef.current.removeEventListener('paste', interceptPaste)
```
Comment (internettrans) [inline src/rich-text-editor.component.js]: the return value of an effect needs to be a function

File: src/rich-text-editor.component.js [hunk 3]
```diff
@@ -27,6 +27,21 @@ export const RichTextEditor = forwardRef((props, editorRef) => {
...
+    }
+  }, [])
```
Comment (internettrans) [inline src/rich-text-editor.component.js]: This should be `[props.pasteFn]`. We should not think of effects as a "mount/unmount" effect, but as a "whenever the inputs change" effect. See https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects for more details

File: src/rich-text-editor.component.js [hunk 4]
```diff
@@ -27,6 +27,21 @@ export const RichTextEditor = forwardRef((props, editorRef) => {
     }
   }
 
+  function interceptPaste (event) {
+    // https://developer.mozilla.org/en-US/docs/Web/Events/paste
+    event.preventDefault()
```
Comment (internettrans) [inline src/rich-text-editor.component.js:32]: By preventing default, the paste won't happen, right? So don't we need to insert the HTML manually to make the paste actually work?

Comment (dvnrsn) [reply src/rich-text-editor.component.js:32]: correct, but the idea here is that the user wants to strip/alter paste perhaps in which case they have to setHTML... or maybe bandicoot just takes the returned value and sets HTML?

```js
function stripHTML(content) {
  ...
}
return (
<RichTextEditor 
  ref={editorRef}
  pasteFn={content => editorRef.current.setHTML(stripHTML(content))

```

Comment (internettrans) [reply src/rich-text-editor.component.js:32]: But pasting shouldn't override the entire rich text editor's content. Only the highlight portion. Right?

File: src/rich-text-editor.component.js [hunk 5]
```diff
@@ -31,11 +31,15 @@ export const RichTextEditor = forwardRef((props, editorRef) => {
     // https://developer.mozilla.org/en-US/docs/Web/Events/paste
     event.preventDefault()
     event.stopPropagation()
-    let paste = (window.clipboardData).getData('text');
+    let paste = (window.clipboardData || event.clipboardData).getData('text/html');
```
Comment (dvnrsn) [inline src/rich-text-editor.component.js:34]: window.clipboardData is for IE it seems, might be able to remove

Comment (dvnrsn) [reply src/rich-text-editor.component.js:34]: https://stackoverflow.com/a/30381888

File: src/rich-text-editor.component.js [hunk 6]
```diff
@@ -31,11 +31,15 @@ export const RichTextEditor = forwardRef((props, editorRef) => {
...
+    console.log('new', paste);
     let newPaste = props.pasteFn(paste)
     const selection = window.getSelection();
     if (!selection.rangeCount) return false;
-    selection.getRangeAt(0).insertNode(document.createTextNode(newPaste));
+    console.log('newPaste', newPaste);
+    const pasteToDOMNode = document.createRange().createContextualFragment(newPaste)
+    console.log('DOM Node', pasteToDOMNode);
+    selection.getRangeAt(0).insertNode(pasteToDOMNode);
```
Comment (dvnrsn) [inline src/rich-text-editor.component.js]: I remove the console logs in next commit

File: src/rich-text-editor.component.js (hunk 5 above)
Comment (dvnrsn) [reply src/rich-text-editor.component.js:34]: NOTICE THE switch from 'text' to 'text/html' here

File: src/rich-text-editor.component.js (hunk 4 above)
Comment (internettrans) [reply src/rich-text-editor.component.js:32]: Have you tested this in Edge? MDN says that it's support is unknown. I just tested it locally and it seemed to work fine, but we'll have to keep an eye on it. https://developer.mozilla.org/en-US/docs/Web/API/Range/createContextualFragment

**Review by internettrans — APPROVED** [review]

**Pull request merged**